### DISCUSSION TITLE
How to use getStaticPaths with a catch-all route nested in a dynamic route?

### DISCUSSION THREAD

**Discussion by Toryn02** [vercel/next.js#27186] [Help] [answered]
Hi there,

I'm attempting to set up a page structure in which I have an optional catch-all route nested within a dynamic route: 

`pages/authors/[author]/[[...book]].js`

However, after several attempts, unfortunately I can't seem to figure out how I should be setting up my getStaticPaths function so that I can use SSG, both for the dynamic [author] routes when no "book" parameter is present in the URL, and for all of the possible [author]/[[...book]] routes as well.  

Using the code below, when I attempt to visit mywebsite.com/authors/orwell/1984, I get the following error:

`Error: A required parameter (book) was not provided as an array in getStaticPaths for /authors/[author]/[[...book]]`

I'm sure that this is user error on my part, but I'm stumped and would very much appreciate any help that folks can provide.  I've included some stripped-down code below to hopefully highlight what I'm looking to accomplish.

```
export default function AuthorBookPage({ author, book }) {
  return (
    <div>
      <h2>{author}</h2>
      <p>{book}</p>
    </div>
  )
}

export async function getStaticProps({ params }) {
  const author = params.author
  const book = params.book ? params.book : 'war-and-peace'

  return { props: { author, book } }
}

export async function getStaticPaths() {
  const paramsArray = [
    { params: { author: 'orwell' } },
    { params: { author: 'cameron' } },
    { params: { author: 'marquez' } },
    { params: { author: 'orwell', book: '1984' } },
    { params: { author: 'cameron', book: 'artists-way' } },
    { params: { author: 'marquez', book: '100-years-of-solitude' } }
  ]

  return {
    paths: paramsArray,
    fallback: false
  }
}
```

Thank you very much in advance!

**Comment by LawJolla** [comment, marked as answer]
Hi @Toryn02 !

I don't know if what you're doing is not possible, but I know it's not standard.

Why not `/authors/[[...catchAll]]` and then...

```
export async function getStaticProps({ params }) {
  const [author, book, ...rest] = params.catchAll
  const book = book ||  'war-and-peace'

  return { props: { author, book } }
}
```

**Reply by Toryn02** [reply]
Thanks very much for taking the time to reply with that suggestion!  I took your idea and re-arranged a few things (using "let" instead of "const" so that I can assign a value to the "book" param if one doesn't exist in the URL route), and this now appears to be working, has the functionality that I'm looking for, and builds without any errors:

```
Page structure: pages/authors/[...author].js

export default function AuthorBookPage({ author, book }) {
  return (
    <div>
      <h2>{author}</h2>
      <p>{book}</p>
    </div>
  )
}

export async function getStaticProps({ params }) {
  let [author, book] = params.author
  if (!book) {
    book = null
  }

  return { props: { author, book } }
}

export async function getStaticPaths() {
  const paramsArray = [
    { params: { author: ['orwell'] } },
    { params: { author: ['cameron'] } },
    { params: { author: ['marquez'] } },
    { params: { author: ['orwell', '1984'] } },
    { params: { author: ['cameron', 'artists-way'] } },
    { params: { author: ['marquez', '100-years-of-solitude'] } }
  ]

  return {
    paths: paramsArray,
    fallback: false
  }
}
```

I'm not sure if there's a cleaner way to confirm whether the second "book" parameter exists in the URL route within my getStaticProps() function, but if anyone has any ideas on how to further clean up the above code, I'd very much welcome the feedback!

**Reply by LawJolla** [reply]
If you're ok with undefined instead of null, you don't need the `!book` check.  The destructured array for book will be `undefined` if not defined.

If you need null, you can optionally set it.
`const [author, book = null] = params.author`

**Reply by Toryn02** [reply]
Great point, and thank you for that!
